File: molecore/src/main/java/io/github/sree/molecore/listeners/WitherSkeletonDeathListener.java
package io.github.sree.molecore.listeners;

import io.github.sree.molecore.enums.Objective;
import io.github.sree.molecore.state.GameState;
import org.bukkit.Material;
import org.bukkit.entity.WitherSkeleton;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.EntityDeathEvent;
import org.bukkit.inventory.ItemStack;

public class WitherSkeletonDeathListener extends GameListener {

    public WitherSkeletonDeathListener(GameState gameState) {
        super(gameState);
    }

    @EventHandler
    public void onWitherSkeletonDeath(EntityDeathEvent event) {
        if (!gameState.isGameStarted()) {
            return;
        }

        if (gameState.getSettings().objective() != Objective.BEACON) {
            return;
        }

        if (!(event.getEntity() instanceof WitherSkeleton)) {
            return;
        }

        if (Math.random() < 0.25) {
            event.getDrops().add(new ItemStack(Material.WITHER_SKELETON_SKULL));
        }
    }
}

File: soulswap/src/main/kotlin/io/github/sree/soulswap/listeners/WitherSkeletonDeathListener.kt
package io.github.sree.soulswap.listeners

import io.github.sree.soulswap.state.GameState
import org.bukkit.Material
import org.bukkit.entity.WitherSkeleton
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDeathEvent
import org.bukkit.inventory.ItemStack

internal class WitherSkeletonDeathListener(val gameState: GameState) : Listener {

    @EventHandler
    fun onWitherSkeletonDeath(event: EntityDeathEvent) {
        if (!gameState.gameStarted || event.entity !is WitherSkeleton) return

        if (Math.random() <= 0.1) event.drops.add(ItemStack(Material.WITHER_SKELETON_SKULL))
    }
}